### ISSUE TITLE
Nan found in hypergradients

### ISSUE THREAD

**Issue by killandy** [lorraine2/implicit-hyper-opt#3]
Recently I have read your paper and find it useful for my work. So I use the hypergradient computing algorithm provided in the code in my method. However, after training few steps, some gradients turn to Nan data. 

Here is my code:
`

    for p in A.aug_parameters():
            if p.grad is not None:
                p.grad = p.grad * 0

    ## train loss: d L_T / d w
    model.train()
    model.zero_grad()

    # data processing
    train_input = A(valid_train_images)

    # train loss
    output, target = model(train_input, valid_train_images)
    train_loss = criterion(output, target)
    d_train_loss_d_w = gather_flat_grad(grad(train_loss, model.encoder_q.parameters(), create_graph=True))

    optimizer_cl.zero_grad()

    ## valid loss
    rand_policy_time = time.time()
    valid_input = RA.aug(valid_images)
    aug_time= time.time()-rand_policy_time

    model.train()
    model.zero_grad()
    valid_output, valid_target = model(valid_input, valid_images)
    valid_loss = criterion(valid_output, valid_target)
    optimizer_cl.zero_grad()
    # d_val_loss_d_theta = gather_flat_grad(grad(valid_loss, model.encoder_q.parameters(), retain_graph=use_reg))
    d_val_loss_d_theta = gather_flat_grad(grad(valid_loss, model.encoder_q.parameters(), retain_graph=False))

    # Initialize the preconditioner and counter
    preconditioner = neumann_hyperstep_preconditioner(d_val_loss_d_theta, d_train_loss_d_w, cur_lr, num_neumann_terms, model)

    # compute d / d lambda (partial Lv / partial w * partial Lt / partial w)
    # = (partial Lv / partial w * partial^2 Lt / (partial w partial lambda))
    direct_grad = 0
    hypergrad = grad(d_train_loss_d_w, A.aug_parameters(), grad_outputs=preconditioner.view(-1))
    indirect_grad = gather_flat_grad(hypergrad)
    hypergrad = direct_grad + indirect_grad
    if check_nan(hypergrad):
        raise Exception('nan')
`

The network is ResNet18, I have tried Adam and SGD optimizers with different lr but still got Nan. Is there any suggestions to avoid this? 

Thanks.
